src/Categories/Category/Monoidal/Interchange.agda

```diff
@@ -0,0 +1,603 @@
+
+    swapInner-assoc : ∀ {X₁ Y₁ Z₁ X₂ Y₂ Z₂} →
+                      [ ((X₁ ⊗₀ X₂) ⊗₀ (Y₁ ⊗₀ Y₂)) ⊗₀ (Z₁ ⊗₀ Z₂) ⇒
+                        (X₁ ⊗₀ (Y₁ ⊗₀ Z₁)) ⊗₀ (X₂ ⊗₀ (Y₂ ⊗₀ Z₂)) ]⟨
+                        i⇒ ⊗₁ id   ⇒⟨ ((X₁ ⊗₀ Y₁) ⊗₀ (X₂ ⊗₀ Y₂)) ⊗₀ (Z₁ ⊗₀ Z₂) ⟩
+                        i⇒         ⇒⟨ ((X₁ ⊗₀ Y₁) ⊗₀ Z₁) ⊗₀ ((X₂ ⊗₀ Y₂) ⊗₀ Z₂) ⟩
+                        α⇒ ⊗₁ α⇒
+                      ≈ α⇒         ⇒⟨ (X₁ ⊗₀ X₂) ⊗₀ ((Y₁ ⊗₀ Y₂) ⊗₀ (Z₁ ⊗₀ Z₂)) ⟩
+                        id ⊗₁ i⇒   ⇒⟨ (X₁ ⊗₀ X₂) ⊗₀ ((Y₁ ⊗₀ Z₁) ⊗₀ (Y₂ ⊗₀ Z₂)) ⟩
+                        i⇒
+                      ⟩
+
+    swapInner-unitˡ : ∀ {X Y} →
+                      [ unit ⊗₀ (X ⊗₀ Y) ⇒ (X ⊗₀ Y) ]⟨
+                        λ⇐ ⊗₁ id   ⇒⟨ (unit ⊗₀ unit) ⊗₀ (X ⊗₀ Y) ⟩
+                        i⇒         ⇒⟨ (unit ⊗₀ X) ⊗₀ (unit ⊗₀ Y) ⟩
+                        λ⇒ ⊗₁ λ⇒
+                      ≈ λ⇒
+                      ⟩
+
+    swapInner-unitʳ : ∀ {X Y} →
+                      [ (X ⊗₀ Y) ⊗₀ unit ⇒ (X ⊗₀ Y) ]⟨
+                        id ⊗₁ λ⇐   ⇒⟨ (X ⊗₀ Y) ⊗₀ (unit ⊗₀ unit) ⟩
+                        i⇒         ⇒⟨ (X ⊗₀ unit) ⊗₀ (Y ⊗₀ unit) ⟩
+                        ρ⇒ ⊗₁ ρ⇒
+                      ≈ ρ⇒
+                      ⟩
+
+  -- The interchange is a natural isomorphism.
+  
+  swapInner-naturalIso : ⊗ ∘F (⊗ ⁂ ⊗) ≃ ⊗ ∘F MonoidalProduct.⊗ M M
+  swapInner-naturalIso = niHelper (record
+    { η       = λ _ → i⇒
+    ; η⁻¹     = λ _ → i⇐
+    ; commute = λ _ → swapInner-natural
+    ; iso     = λ _ → swapInner.iso
+    })
+
+-- Shorthands for composing and inverting isomorphisms.
+
+open ≅ using () renaming (refl to idᵢ; sym to _⁻¹)
+infixr 9 _∘ᵢ_
+_∘ᵢ_ = λ {A B C} f g → ≅.trans {A} {B} {C} g f
+
+-- Braided monoidal categories have an interchange map.
+
+module BraidedInterchange {M : Monoidal C} (B : Braided M) where
+  open MonoidalReasoning M
+  open MonoidalUtilities M
+  open Braided B renaming (associator to α)
+  open Shorthands
+
+  -- Shorthands for braiding.
+
+  σ : ∀ {X Y} → X ⊗₀ Y ≅ Y ⊗₀ X
+  σ = braiding.FX≅GX 
+  module σ  {X Y} = _≅_ (σ {X} {Y})
+  private
+    σ⇒ = σ.from
+    σ⇐ = σ.to
+
+  -- The "four middle interchange" for braided tensor products.
+
+  swapInner : ∀ {W X Y Z} → (W ⊗₀ X) ⊗₀ (Y ⊗₀ Z) ≅ (W ⊗₀ Y) ⊗₀ (X ⊗₀ Z)
+  swapInner = α ⁻¹ ∘ᵢ idᵢ ⊗ᵢ (α ∘ᵢ σ ⊗ᵢ idᵢ ∘ᵢ α ⁻¹) ∘ᵢ α
+
+  module swapInner {W X Y Z} = _≅_ (swapInner {W} {X} {Y} {Z})
+  private
+    i⇒ = swapInner.from
+    i⇐ = swapInner.to
+
+  -- The interchange is a natural isomorphism.
+
+  swapInner-natural : ∀ {W₁ X₁ Y₁ Z₁ W₂ X₂ Y₂ Z₂}
+                      {f : W₁ ⇒ W₂} {g : X₁ ⇒ X₂} {h : Y₁ ⇒ Y₂} {i : Z₁ ⇒ Z₂} →
+                      i⇒ ∘ (f ⊗₁ g) ⊗₁ (h ⊗₁ i) ≈ (f ⊗₁ h) ⊗₁ (g ⊗₁ i) ∘ i⇒
+  swapInner-natural {f = f} {g} {h} {i} = begin
+      (α⇐ ∘ id ⊗₁ (α⇒ ∘ σ⇒ ⊗₁ id ∘ α⇐) ∘ α⇒) ∘ (f ⊗₁ g) ⊗₁ (h ⊗₁ i)
+    ≈⟨ pullʳ (pullʳ assoc-commute-from) ⟩
+      α⇐ ∘ id ⊗₁ (α⇒ ∘ σ⇒ ⊗₁ id ∘ α⇐) ∘ f ⊗₁ g ⊗₁ (h ⊗₁ i) ∘ α⇒
+    ≈⟨ refl⟩∘⟨ extendʳ (parallel id-comm-sym (begin
```

Comment:
Here I'm not sure what you mean by 'inner lemmas'. What do you think should be factored out, the nested equational proof or the 'pseudo-pattern'?

Assuming you mean the nested equations, it's interesting that you think of those as 'inner lemmas'. I don't think of these as lemmas, but as equations on subexpressions. I use them when I notice that there's a series of steps in an equational proof that only change a subexpression `a = b = ... = c` of some large and complicated expression `C[a] = C[b] = ... = C[c]`. In such cases there's a lot of unnecessary (and distracting) repetition and I tend to open a nested proof. I can see how this suggests that there are hidden 'inner lemmas' to be extracted. You're probably right for some of those, but for others I think that's a red herring. Take, for example, the first nested proof in `swapInner-assoc` which is of the form `α⇒ ⊗₁ α⇒ = <complicated expression>`. You may wonder, why did I choose `<complicated expression>`? The answer is: it's a garbage expression that accumulated during the rest of the proof and I had to resolve it. It's just a tangle of associators that we know must be equal to `α⇒ ⊗₁ α⇒` because of the coherence theorem for monoidal categories. If we had a proof of that theorem, I'd just have used that. Sure, it's a lemma in some sense, but then so is `2 a = (a b + b^2)/b + a - b`. That being said, there might be some actual 'lemmas' in there that are worth extracting, I'm not sure.

I'd like to propose the following: why don't we keep the proofs as they are for the PR and I'll keep the 'interchange' branch around for refactoring/experimentation. I suspect that improving these proofs will involve a non-trivial amount of backtracking and I don't think doing this in a PR is a good idea. If the work is done in a feature branch everyone can check it out, play with it and suggest concrete changes. And we can discuss possible strategies on in a Zulip branch. Of course I'll still fix naming and documentation, etc. as you requested.

WDYT?

Reply:
I fully agree that they are 'but as equations on subexpressions'. I just happen to also make these into (private) lemmas, for readability's sake.  Whether there are some lemmas that should be made public because they can be re-used, that's a different thing. I have not had a chance to look deeply enough to see patterns.

I agree that we can do this as a later phase.